### PR TITLE
Move Mul node before Reshape in GroupConvolutionMultiplyFusion

### PR DESCRIPTION
**Pull request by mateusztabaka**
In following scenario:

                                    +-----------+
                                    |  weights  |
                                    +-----------+
                                         |
                                         v
                                    +-----------+
                                    |  Reshape  |
                       ....         +-----------+
                          \              /
                           \            /
                            v          v
                       +--------------------+
                       |  GroupConvolution  |
                       +--------------------+
                                |
                                v
                           +-----------+      +----------+
                           |  Multiply | <--- | Constant |
                           +-----------+      +----------+

if a second input (weights) to GroupConvolution is a Reshape,
we should apply Multiply node directly to weights rather than Reshape node:

                                    +-----------+
                                    |  weights  |
                                    +-----------+
                                         |
                                         v
                                    +-----------+      +----------+
                                    |  Multiply | <--- | Constant |
                                    +-----------+      +----------+
                                         |
                                         v
                                    +-----------+
                                    |  Reshape  |
                       ....         +-----------+
                          \              /
                           \            /
                            v          v
                       +--------------------+
                       |  GroupConvolution  |
                       +--------------------+
                                |
                                v

That approach has no side effects in the usual scenario when weights are constant,
but it's necessary for models with FakeQuantize since it gives possibility to perform
GroupConvolutionTransformation from LPT transformations.

Ticket: 44783

### CODE REVIEW

File: inference-engine/src/transformations/src/transformations/common_optimizations/conv_mul_fusion.cpp [hunk 1]
```diff
@@ -118,27 +119,49 @@ ngraph::pass::GroupConvolutionMultiplyFusion::GroupConvolutionMultiplyFusion() {
...
+            new_conv = m_conv->copy_with_new_inputs({m_input, new_weights});
+            copy_runtime_info({m_conv, m_mul}, {final_const.get_node_shared_ptr(), new_weights, new_conv});
+        }
 
         // Replace Convolution->Multiply with Convolution with new inputs
-        auto new_conv = m_conv->copy_with_new_inputs({m_input, weights_multiply});
         new_conv->set_friendly_name(m_mul->get_friendly_name());
-        copy_runtime_info({m_conv, m_mul}, {new_conv, final_const.get_node_shared_ptr(), weights_multiply});
         replace_node(m_mul, new_conv);
+
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/conv_mul_fusion.cpp:164]: I would simplify this code like this:
```cpp
auto new_weights = std::make_shared<opset4::Multiply>(m_weights, final_const);
if (are_weights_reshaped) {
    reshape->input(0).replace_source_output(new_weights);
} else {
    m_conv->input(1).replace_source_output(new_weights);
}
m_conv->set_friendly_name(m_mul->get_friendly_name());
m_mul->output(0).replace(m_conv->output(0));
copy_runtime_info({m_mul}, {m_conv, new_weights});
return true;
```

**Review by GlebKazantaev — CHANGES_REQUESTED** [review]

File: inference-engine/src/transformations/src/transformations/common_optimizations/conv_mul_fusion.cpp [hunk 2]
```diff
@@ -93,15 +93,15 @@ ngraph::pass::GroupConvolutionMultiplyFusion::GroupConvolutionMultiplyFusion() {
...
         const auto & m_const = pattern_to_output.at(mul_const);
-        const auto & m_input = pattern_to_output.at(input);
         const auto & m_conv = pattern_to_output.at(conv).get_node_shared_ptr();
         const auto & m_mul = pattern_to_output.at(mul).get_node_shared_ptr();
 
-        const auto & G = m_weights.get_partial_shape()[0].get_length();
-        const auto & O = m_weights.get_partial_shape()[1].get_length();
-        const auto & weights_rank = m_weights.get_partial_shape().rank().get_length();
+        const auto & weights_shape = m_weights.get_partial_shape();
+        const auto & G = weights_shape[0].get_length();
```
Comment (GlebKazantaev) [inline inference-engine/src/transformations/src/transformations/common_optimizations/conv_mul_fusion.cpp:102]: Please add `pattern::has_static_dims({0, 1})` to the weights pattern predicate.

**Review by GlebKazantaev — APPROVED** [review]
LGTM from transformation perspective. Waiting for @eshoguli review.

**Comment by jane-intel** [comment]
Don't we have such a Reshape eliminator already [here](https://github.com/openvinotoolkit/openvino/blob/master/inference-engine/src/transformations/src/transformations/common_optimizations/fq_reshape_fusion.cpp):

Introduced during: https://github.com/openvinotoolkit/openvino/pull/2388

File: inference-engine/src/transformations/src/transformations/common_optimizations/conv_mul_fusion.cpp [hunk 3]
```diff
@@ -85,23 +85,23 @@ NGRAPH_RTTI_DEFINITION(ngraph::pass::GroupConvolutionMultiplyFusion, "GroupConvo
...
```
Comment (eshoguli) [inline inference-engine/src/transformations/src/transformations/common_optimizations/conv_mul_fusion.cpp:102]: Should we check if group count (`G`) equals to channels count and return `false` if not?

**Review by eshoguli — COMMENTED** [review]
Possible issue from LPT side: 

Issue #1. Operations on weights can be changed. LPT support predefined patterns on weights (GroupConvolution as example):

**FQ on weights**
```
 [activation]   FakeQuantize
     \          /
      \       Reshape
       \      /
    GroupConvolution
```
**New weights format**
```
              Constant
               |
              Convert  Constant
               |      /
              Subtract Constant
               |      /
              Multiply
               |
[activation]  Reshape
      \       /
    GroupConvolution
```

Note, please, there are no duplicated `Multiply` operations here. 
**Can you confirm that operations on weights will be not changed for LPT (next transformations will fuse new `Multiply` to exist `FakeQuantize` or exist `Multiply`)?**

Issue #2. In this case LPT team adds CPU & GPU regression plugin tests to check whole pipeline. I would be grateful if you could add this CPU & GPU regression test. I think it can do later but create task for that.

File: inference-engine/tests/functional/inference_engine/transformations/conv_fusion_test.cpp [hunk 4]
```diff
@@ -168,3 +170,165 @@ INSTANTIATE_TEST_CASE_P(DISABLED_ConvMulFusionNegative, ConvFusionTests,
...
+    auto res = compare_functions(f, f_ref);
+    ASSERT_TRUE(res.first) << res.second;
+}
+
+
+TEST(GroupConvMulFusion, WeightsWithFakeQuantizeAndReshape) {
+    std::shared_ptr<Function> f(nullptr), f_ref(nullptr);
+    {
+        auto data = std::make_shared<opset5::Parameter>(element::f32, Shape{1, 4, 7, 7});
+        auto weights = opset5::Constant::create(element::f32, Shape{4, 1, 3, 3}, std::vector<float>(36, 1));
```
Comment (GlebKazantaev) [inline inference-engine/tests/functional/inference_engine/transformations/conv_fusion_test.cpp:299]: Could you please add tests where the `O` dim is greater than 1?

**Comment by GlebKazantaev** [comment]
In general LGTM.

**Review by ilyachur — APPROVED** [review]

**Comment by ilyachur** [comment]
@eshoguli Please have a look

**Review by postrational — APPROVED** [review]

**Review by eshoguli — APPROVED** [review]
Great tests for LPT GroupConvolution transformation, thanks!

**Pull request merged**